File: tests/r_parity/31_dfl.R
# StatsPAI DFL reweighting parity (R side) -- Module 31.
#
# Reads data/31_dfl.csv and runs ddecompose::dfl_decompose.
# Tolerance: rel < 1e-3 on the gap, composition, and structure.
# ddecompose reference_0=TRUE reweights group 0; the Python side uses
# StatsPAI reference=1 because StatsPAI reference indexes the target
# covariate distribution.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(ddecompose)
})

MODULE <- "31_dfl"

df <- read_csv_strict(MODULE)

# ddecompose's dfl_decompose: formula uses `outcome ~ covariates | group`
# syntax via Formula package. Group is passed as a bare symbol.
df$female <- as.factor(df$female)

fit <- ddecompose::dfl_decompose(
  formula = log_wage ~ educ + exper,
  data = df,
  group = female,
  reference_0 = TRUE,
  bootstrap = FALSE,
  statistics = c("mean")
)

# fit$decomposition_other_statistics is a data.frame with rows
# per statistic and columns: 'Observed difference', 'Composition
# effect', 'Structure effect'. ddecompose returns the (group=0) -
# (group=1) gap when reference_0=TRUE; sp returns the
# (group=0_mean) - (group=1_mean) gap with the same sign convention.
md <- fit$decomposition_other_statistics
gap_est   <- md[md$statistic == "mean", "Observed difference"][1]
comp_est  <- md[md$statistic == "mean", "Composition effect"][1]
struc_est <- md[md$statistic == "mean", "Structure effect"][1]

# Note: ddecompose's "Observed difference" is reference_minus_other,
# which equals -(sp's gap) when reference_0=TRUE because sp
# computes (group_b_mean - group_a_mean) by convention. Flip sign
# for direct comparison.
gap_est   <- -gap_est
comp_est  <- -comp_est
struc_est <- -struc_est

rows <- list(
  parity_row(MODULE, "gap",         estimate = gap_est,    n = nrow(df)),
  parity_row(MODULE, "composition", estimate = comp_est,   n = nrow(df)),
  parity_row(MODULE, "structure",   estimate = struc_est,  n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(method = "ddecompose::dfl_decompose",
                           reference_0 = TRUE,
                           statistic = "mean",
                           sign_flipped = TRUE))

File: tests/r_parity/31_dfl.py
"""StatsPAI DiNardo-Fortin-Lemieux reweighting parity (Python side).
Module 31.

Generates the same wage-gap dataset as Module 30 but with N=1600
to give DFL kernel-density estimation a stable signal. The companion
31_dfl.R uses ddecompose::dfl_decompose.

Tolerance: rel < 1e-3 on the DFL gap, composition, and structure.
``ddecompose::dfl_decompose(reference_0=TRUE)`` reweights group 0 to
match group 1. StatsPAI's ``reference`` indexes the target covariate
distribution, so the matching counterfactual is ``reference=1``.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "31_dfl"


def make_data(n_per: int = 800, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    female = np.repeat([0, 1], n_per)
    educ = rng.normal(15, 3, 2 * n_per)
    exper = rng.normal(10, 5, 2 * n_per) - 1.0 * (female == 1)
    log_wage = (
        2.0 + 0.07 * educ + 0.02 * exper - 0.20 * female
        + rng.normal(0, 0.3, 2 * n_per)
    )
    return pd.DataFrame({
        "log_wage": log_wage, "educ": educ, "exper": exper,
        "female": female,
    })


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    res = sp.decompose(
        "dfl", data=df, y="log_wage", group="female",
        x=["educ", "exper"], reference=1,
    )

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "gap",
                     estimate=float(res.gap), n=int(len(df))),
        ParityRecord(MODULE, "py", "composition",
                     estimate=float(res.composition), n=int(len(df))),
        ParityRecord(MODULE, "py", "structure",
                     estimate=float(res.structure), n=int(len(df))),
        ParityRecord(MODULE, "py", "stat_a",
                     estimate=float(res.stat_a), n=int(len(df))),
        ParityRecord(MODULE, "py", "stat_b",
                     estimate=float(res.stat_b), n=int(len(df))),
        ParityRecord(MODULE, "py", "stat_cf",
                     estimate=float(res.stat_cf), n=int(len(df))),
    ]

    write_results(
        MODULE, "py", rows,
        extra={
            "reference": int(res.reference),
            "stat": str(res.stat),
            "ddecompose_reference_mapping": (
                "StatsPAI reference=1 matches ddecompose "
                "reference_0=TRUE because StatsPAI reference indexes "
                "the target covariate distribution."
            ),
        },
    )


if __name__ == "__main__":
    main()
